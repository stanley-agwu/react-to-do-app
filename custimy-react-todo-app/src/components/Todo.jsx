import React, { useEffect, useState } from 'react';

import TodoItem from './TodoItem';
import AddItem from './AddItem';
import { StyledTodo, StyledTodoList, StyledFooter } from './styled';

const Todo = () => {
  const [todoItems, setTodoItems] = useState([]);

  useEffect(() => {
    // Load from storage
    if (typeof Storage !== 'undefined') {
      // localStorage supported.
      const todos = JSON.parse(localStorage.getItem('todoItems'));
      if (todos) {
        setTodoItems(todos);
      }
    } else {
      // Using cookies here :(
      let todos = document.cookie;
      if (todos !== '') {
        todos = todos.split('; ');
        const decodedTodos = [];
        todos.forEach((item) => {
          const splitItem = item.split('=');
          decodedTodos.push({
            date: splitItem[0],
            todo: splitItem[1],
          });
        });
        setTodoItems(decodedTodos);
      }
    }
  }, []);

  useEffect(() => {
    // save to storage
    if (typeof Storage !== 'undefined') {
      // localStorage supported.
      localStorage.setItem('todoItems', JSON.stringify(todoItems));
    } else {
      // Using cookies here :(
      const todos = todoItems;
      todos.forEach((item) => {
        document.cookie = `${item.date}=${item.todo}; expires=Fri, 31 Dec 9999 23:59:59 GMT`;
      });
    }
  }, [todoItems]);

  const addItem = (todoText, dueDate, priority) => {
    const todos = todoItems;
    todos.push({ todo: todoText, dueDate, priority, date: new Date().getTime() });
    setTodoItems([...todos]);
  };

  const deleteItem = (item) => {
    const todos = todoItems.filter((todoItem) => item !== todoItem);
    setTodoItems(todos);
  };

  const editItem = (item, value) => {
    const todos = todoItems;
    const editTodo = todos.find((todoItem) => todoItem.date === item.date);
    editTodo.todo = value;
    setTodoItems([...todos]);
  };

  return (
    <StyledTodo className="wrapper">
      <h1>todos.</h1>
      <AddItem addItem={addItem} />
      <StyledTodoList>
        <ul>
          {todoItems
            .sort((a, b) => (b.date - a.date))
            .map((item) => (
              <TodoItem
                item={item}
                key={item.date}
                deleteItem={deleteItem}
                editItem={editItem}
              />
            ))}
        </ul>
      </StyledTodoList>
      <StyledFooter>
        <p>A simple Todo app made with React.</p>
        <p>
          It&#39;s ok to close the browser, all items will be saved locally.
        </p>
        <p>
          If you use Private browsing or Incognito mode, the app will be unable
          to save changes.
        </p>
      </StyledFooter>
    </StyledTodo>
  );
};

export default Todo;
