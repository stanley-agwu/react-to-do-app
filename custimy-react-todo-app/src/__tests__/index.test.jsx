import React from 'react';
import { render } from '@testing-library/react';
import Todo from '../components/Todo';

test('renders without throwing errors', () => {
  const container = render(<Todo />);
  expect(container).toMatchSnapshot();
});
