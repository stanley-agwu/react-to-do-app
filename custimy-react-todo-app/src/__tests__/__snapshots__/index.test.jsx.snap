// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without throwing errors 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="styled__StyledTodo-lucpj2-2 dbtvUP wrapper"
      >
        <h1>
          todos.
        </h1>
        <form
          class="styled__StyledAddItem-lucpj2-0 jtAjRE"
        >
          <input
            placeholder="New todo..."
            required=""
            type="text"
            value=""
          />
          <div>
            <input
              required=""
              type="date"
              value=""
            />
            <select
              required=""
            >
              <option
                value=""
              >
                Select Priority
              </option>
              <option
                value="URGENT"
              >
                URGENT
              </option>
              <option
                value="IMPORTANT"
              >
                IMPORTANT
              </option>
              <option
                value="SECONDARY"
              >
                SECONDARY
              </option>
              <option
                value="PRIMARY"
              >
                PRIMARY
              </option>
            </select>
          </div>
          <button
            type="submit"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-w-14 fa-plus"
              data-icon="plus"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 448 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
                fill="currentColor"
              />
            </svg>
          </button>
        </form>
        <div
          class="styled__StyledTodoList-lucpj2-3 zklyv"
        >
          <ul />
        </div>
        <div
          class="styled__StyledFooter-lucpj2-4 gMOygU"
        >
          <p>
            A simple Todo app made with React.
          </p>
          <p>
            It's ok to close the browser, all items will be saved locally.
          </p>
          <p>
            If you use Private browsing or Incognito mode, the app will be unable to save changes.
          </p>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="styled__StyledTodo-lucpj2-2 dbtvUP wrapper"
    >
      <h1>
        todos.
      </h1>
      <form
        class="styled__StyledAddItem-lucpj2-0 jtAjRE"
      >
        <input
          placeholder="New todo..."
          required=""
          type="text"
          value=""
        />
        <div>
          <input
            required=""
            type="date"
            value=""
          />
          <select
            required=""
          >
            <option
              value=""
            >
              Select Priority
            </option>
            <option
              value="URGENT"
            >
              URGENT
            </option>
            <option
              value="IMPORTANT"
            >
              IMPORTANT
            </option>
            <option
              value="SECONDARY"
            >
              SECONDARY
            </option>
            <option
              value="PRIMARY"
            >
              PRIMARY
            </option>
          </select>
        </div>
        <button
          type="submit"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-w-14 fa-plus"
            data-icon="plus"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 448 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
              fill="currentColor"
            />
          </svg>
        </button>
      </form>
      <div
        class="styled__StyledTodoList-lucpj2-3 zklyv"
      >
        <ul />
      </div>
      <div
        class="styled__StyledFooter-lucpj2-4 gMOygU"
      >
        <p>
          A simple Todo app made with React.
        </p>
        <p>
          It's ok to close the browser, all items will be saved locally.
        </p>
        <p>
          If you use Private browsing or Incognito mode, the app will be unable to save changes.
        </p>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
